feat: blobs 2.

MirandaWood opened the pull request:
Electric boogaloo (cont. of #9302)

This reverts commit a415f6552ae9893699747b4d1fc799553e9a9a7e.

MirandaWood committed:
Revert "fix: Revert "feat: blobs. (#9302)" (#10187)"

This reverts commit a415f6552ae9893699747b4d1fc799553e9a9a7e.

ludamad commented:
BTW add label 'network-all' when ready

MirandaWood committed:
chore: comments/renaming in Rollup.sol

MirandaWood committed:
feat: give up and make blob oracle for simulation

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
fix: post merge fixes, fmt, fixtures

MirandaWood commented on yarn-project/ethereum/src/test/start_anvil.ts:
```diff
@@ -22,6 +22,7 @@ export async function startAnvil(l1BlockTime?: number): Promise<{ anvil: Anvil;
         anvilBinary,
         port: ethereumHostPort,
         blockTime: l1BlockTime,
+        codeSizeLimit: 0x7000,
```
Rollup.sol now very large and overflows contract size limits :( will try and bring it down in future commits

MirandaWood committed:
chore: dummy change to fix ci runner

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
fix: do horrible things to reduce contract size

MirandaWood commented on l1-contracts/src/core/libraries/HeaderLib.sol:
```diff
@@ -110,10 +110,11 @@ library HeaderLib {
    * @return The decoded header
    */
-  function decode(bytes calldata _header) internal pure returns (Header memory) {
+  function decode(bytes calldata _header) public pure returns (Header memory) {
```
Did some pretty nasty things to bring down the contract size in this commit (64586d8b8760157d2821045557af5a2111ed9552), will look over again once tests are passing.

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: post merge fixes, add libs to gen contracts, comments

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
feat: integrate private logs, add blob oracle issue num, regen fixtures, constants

MirandaWood commented on yarn-project/noir-protocol-circuits-types/src/index.ts:
```diff
@@ -776,28 +776,33 @@ export function foreignCallHandler(name: string, args: ForeignCallInput[]): Prom
         .flat()
         .map((field: string) => fromACVMField(field)),
     );
+    const blobsAsFr = paddedBlobsAsFr.slice(0, spongeBlob.expectedFields);
+    // NB: the above used to be:
+    // const blobsAsFr: Fr[] = args[0].map((field: string) => fromACVMField(field)).filter(field => !field.isZero());
+    // ...but we now have private logs which have a fixed number of fields and may have 0 values.
+    // TODO(Miranda): trim 0 fields from private logs
```
Now we have private logs which are 18 fields each, when the log doesn't use all 18 fields we introduce 0s in the blob. The above works for now but I will trim 0s and add a nested length prefix for each log once compile times go back down a bit.

MirandaWood committed:
chore: update proof fixture

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood reviewed (commented):
Note: The timeout issues are resolved here: https://github.com/AztecProtocol/aztec-packages/pull/10341
I'm trying to see if I can actually merge that PR before resolving conflicts here.

MirandaWood committed:
fix: revert merge changes, compile time down from 90 to 10 mins (#10341)

Reverted the noir changes from this merge:
https://github.com/AztecProtocol/aztec-packages/commit/fe6fa6888f25a544e4ee5b202540f963fae0b913#diff-44fc1bda823fb66518af03520024bce3369ceb8114807d5ca9cf301d8bfcec7cL563
Bootstrapping protocol circuits went from 90mins to 10mins at the
compile stage.
(Apologies for lack of formatting)

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: post merge fixes, fmt

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: post merge fixes + update proof fixture

MirandaWood committed:
chore: some orch cleanup

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: remake consts after merge

iAmMichaelConnor reviewed (approved):
Approved, in the hope that this can be merged soon :)
... with known issues that can be addressed in smaller PRs thereafter (e.g. trickery to get circuit times down, and some Solidity contract size limit reshuffling).

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
feat: redo blob publishing

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
feat: rearrange all blob l1 logic, replicate in ts

MirandaWood committed:
feat: add gas estimation for blob txs, cleanup contracts, increase helm timeout

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood commented on yarn-project/sequencer-client/src/publisher/l1-publisher.ts:
```diff
@@ -170,6 +170,9 @@ export class L1Publisher {
   protected account: PrivateKeyAccount;
   protected ethereumSlotDuration: bigint;
 
+  // @note - with blobs, the below estimate seems too large.
+  // Total used for full block from int_l1_pub e2e test: 1m (of which 86k is 1x blob)
+  // Total used for emptier block from above test: 429k (of which 84k is 1x blob)
```
@ TMNT team - I didn't change the propose guess from 12m to something closer to reality, as it may be super high to account for the worst case?
Using `integration_l1_publisher.test.ts`, I saw:

- blocks with bloated txs used 1m gas (of which 86k = 1x blob)
- blocks with empty txs used 430k gas (of which 84k = 1x blob)

I also added blob gas estimation below (`blobEvaluationGas`). The contract doesn't waste gas on evaulating empty blobs (if we try and cheat and only evaluate some of 3 used blobs, the header check and epoch proof checks will fail).

MirandaWood reviewed (commented):
Note: the kind network tests are all passing locally, but not on CI for some reason.

MirandaWood committed:
chore: typos, comments clarifications

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: fix bad noir merge formatting

MirandaWood committed:
fix: change slot overrides after merge

MirandaWood committed:
chore: update timeouts, comment, grab custom error from ext lib

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: post merge formatting

MirandaWood committed:
chore: add leolib abi to error handling

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: inc test timeout after sync

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: inc timeout

MirandaWood committed:
chore: increase! those! timeouts!

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: header -> block_header to match merge

MirandaWood committed:
chore: guess what.

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: fixture + comment out slower test

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: remove some warnings from blob code

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: post merge formatting

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: remake constants after merge

MirandaWood committed:
fix: bad merge

MirandaWood committed:
chore: proof fixture

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

TomAFrench committed:
feat: reduce the number unconstrained boundary hops in bloblib (#10603)

I've taken a quick look through the bloblib to see if we can make any
optimizations on the Noir code itself. This PR has a few changes from
this:

- I've removed the usage of `std::as_witness` as all the usage was
applied to the return values from unconstrained functions (and so will
be witnesses anyway).
- This shouldn't have affected performance but it also wasn't providing
us with any benefits.
- Rather than calling many unconstrained functions from a constrained
function I have created unconstrained helpers which will perform all
calculations and then return any values which are necessary for applying
constraints.
- This minimises the amount of data which gets pushed from the ACVM to
the Brillig VM and vice versa. This is important as we no longer need to
read all of arguments from calldata and cast them to the correct types
or arrange the results in memory to be read back into the main witness
map by the ACVM.
- This means that for the calculation of a single `partial_sum`, we
would have to spin up and tear down the brillig VM 17(!) times whereas
now it will be done just once.
- I've removed some cases were we'd use
`BigNum::evaluate_quadratic_expression` in order to perform simple
arithmetic and replaced them with the equivalent constrained operation.
An example for this is calculating `z_pow_d`.

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
fix: add temp brillig skip to run CI

MirandaWood committed:
chore: update prover.tomls, ensure block-root is populated

MirandaWood committed:
chore: add brillig gates skip to earthfile

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
chore: reenable test

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/blob-circuit-2
